<div align="right" style=" font-size: 80%; text-align: center; margin: 0 auto">
<img src="https://raw.githubusercontent.com/Explore-AI/Pictures/master/ExploreAI_logos/Logo blue_dark.png"  style="width:25px" align="right";/>
</div>

# Using value-based window functions
© ExploreAI Academy

In this notebook, we will explore the use of value-based window functions to access values from the previous row and use these values to calculate the rate of change between consecutive rows.

> ⚠️ This notebook will not run on Google Colab because it cannot connect to a local database. Please make sure that this notebook is running on the same local machine as your MySQL Workbench installation and MySQL `united_nations` database.

## Learning objectives

In this train, we will learn:
- How to use the `LAG()` value-based window function to extract particular column values from the previous row.
- How the results from `LAG()` can be used to perform analysis such as calculating the rate of change between consecutive values. 
 

## Overview

Say we want to investigate how the percentage of managed drinking water changes from one year to the next in every country. We can add a new column with the previous year's percentage of managed drinking water using the `LAG()` function. This is a value-based window function that extracts the value of a specific column from a previous row. 

## Connecting to our MySQL database

We will use our `Access_to_Basic_Services` table in our `united_nations` database we created in MySQL Workbench. We can apply the same queries we used in MySQL Workbench in this notebook if we connect to our MySQL server by running the cells below.


In [1]:
# Load and activate the SQL extension to allow us to execute SQL in a Jupyter notebook. 
# If you get an error here, make sure that mysql and pymysql are installed correctly. 

%load_ext sql

In [2]:
# Establish a connection to the local database using the '%sql' magic command.
# Replace 'password' with our connection password and `db_name` with our database name. 
# If you get an error here, please make sure the database name or password is correct.

%sql mysql+pymysql://root:password@localhost:3306/united_nations

## Exercise

Let us enter the following base query which selects the three columns we will be using from our `Access_to_Basic_Services` table: `Country_name`, `Time_period`, and `Pct_managed_drinking_water_services`. 


In [4]:
%%sql

SELECT
    Country_name,
    Time_period,
    Pct_managed_drinking_water_services
FROM 
    united_nations.Access_to_Basic_Services
LIMIT 10  

 * mysql+pymysql://root:***@localhost:3306/united_nations
10 rows affected.


[('Kazakhstan', 2015, Decimal('94.67')),
 ('Kazakhstan', 2016, Decimal('94.67')),
 ('Kazakhstan', 2017, Decimal('95.00')),
 ('Kazakhstan', 2018, Decimal('95.00')),
 ('Kazakhstan', 2019, Decimal('95.00')),
 ('Kazakhstan', 2020, Decimal('95.00')),
 ('Kyrgyzstan', 2015, Decimal('89.67')),
 ('Kyrgyzstan', 2016, Decimal('90.33')),
 ('Kyrgyzstan', 2017, Decimal('91.00')),
 ('Kyrgyzstan', 2018, Decimal('91.33'))]

### 1. Add a new column with the previous year's percentage of managed drinking water.

Add the line with the `LAG()` function to the base query above to extract the previous year's percentage of managed drinking water within each country. Store the results in a new column.

In [11]:
%%sql

SELECT
    Country_name,
    Time_period,
    Pct_managed_drinking_water_services,
    LAG (Pct_managed_drinking_water_services) OVER (
        PARTITION BY
            Country_name
        ORDER BY
            Time_period ASC
    ) AS Prev_year_pct_managed_drinking_water_services
FROM
    united_nations.Access_to_Basic_Services
LIMIT
    10

 * mysql+pymysql://root:***@localhost:3306/united_nations
10 rows affected.


[('Afghanistan', 2015, Decimal('67.00'), None),
 ('Afghanistan', 2016, Decimal('69.67'), Decimal('67.00')),
 ('Afghanistan', 2017, Decimal('72.33'), Decimal('69.67')),
 ('Afghanistan', 2018, Decimal('75.33'), Decimal('72.33')),
 ('Afghanistan', 2019, Decimal('78.00'), Decimal('75.33')),
 ('Afghanistan', 2020, Decimal('80.33'), Decimal('78.00')),
 ('Algeria', 2015, Decimal('92.00'), None),
 ('Algeria', 2016, Decimal('93.00'), Decimal('92.00')),
 ('Algeria', 2017, Decimal('93.00'), Decimal('93.00')),
 ('Algeria', 2018, Decimal('93.00'), Decimal('93.00'))]

### 2. Determine the Annual Rate of Change between consecutive years.

Adding on to the query above, let us go further and determine the Annual Rate of Change between consecutive years found by calculating the difference between a measurement and its lag.

In [15]:
%%sql
SELECT
    Country_name,
    Time_period,
    Pct_managed_drinking_water_services,
    LAG (Pct_managed_drinking_water_services) OVER (
        PARTITION BY
            Country_name
        ORDER BY
            Time_period ASC
    ) AS Prev_year_pct_managed_drinking_water_services,
    Pct_managed_drinking_water_services - LAG (Pct_managed_drinking_water_services) OVER (
        PARTITION BY
            Country_name
        ORDER BY
            Time_period ASC
    ) AS ARC_pct_managed_drinking_water_services
FROM
    united_nations.Access_to_Basic_Services
LIMIT
    10

 * mysql+pymysql://root:***@localhost:3306/united_nations
10 rows affected.


[('Afghanistan', 2015, Decimal('67.00'), None, None),
 ('Afghanistan', 2016, Decimal('69.67'), Decimal('67.00'), Decimal('2.67')),
 ('Afghanistan', 2017, Decimal('72.33'), Decimal('69.67'), Decimal('2.66')),
 ('Afghanistan', 2018, Decimal('75.33'), Decimal('72.33'), Decimal('3.00')),
 ('Afghanistan', 2019, Decimal('78.00'), Decimal('75.33'), Decimal('2.67')),
 ('Afghanistan', 2020, Decimal('80.33'), Decimal('78.00'), Decimal('2.33')),
 ('Algeria', 2015, Decimal('92.00'), None, None),
 ('Algeria', 2016, Decimal('93.00'), Decimal('92.00'), Decimal('1.00')),
 ('Algeria', 2017, Decimal('93.00'), Decimal('93.00'), Decimal('0.00')),
 ('Algeria', 2018, Decimal('93.00'), Decimal('93.00'), Decimal('0.00'))]

## Solutions

### 1. Add a new column with the previous year's percentage of managed drinking water.

We apply the `LAG()` function as follows:

Firstly, this will partition our dataset by country, that is, the `Country_name` column, then order each partition by year, that is, the `Time_period` column, from the lowest to the highest. 

Then, the `LAG()` function will be used to extract the previous year's percentage of managed drinking water within a country partition and based on the resulting order of rows from the `ORDER BY` clause. The results will then be stored in a new column `Prev_year_pct_managed_drinking_water_services`.

In [12]:
%%sql

SELECT
    Country_name,
    Time_period,
    Pct_managed_drinking_water_services,
    LAG(Pct_managed_drinking_water_services) OVER( PARTITION BY Country_name 
    ORDER BY Time_period ASC) AS Prev_year_pct_managed_drinking_water_services
FROM 
    united_nations.Access_to_Basic_Services

LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/united_nations
10 rows affected.


[('Afghanistan', 2015, Decimal('67.00'), None),
 ('Afghanistan', 2016, Decimal('69.67'), Decimal('67.00')),
 ('Afghanistan', 2017, Decimal('72.33'), Decimal('69.67')),
 ('Afghanistan', 2018, Decimal('75.33'), Decimal('72.33')),
 ('Afghanistan', 2019, Decimal('78.00'), Decimal('75.33')),
 ('Afghanistan', 2020, Decimal('80.33'), Decimal('78.00')),
 ('Algeria', 2015, Decimal('92.00'), None),
 ('Algeria', 2016, Decimal('93.00'), Decimal('92.00')),
 ('Algeria', 2017, Decimal('93.00'), Decimal('93.00')),
 ('Algeria', 2018, Decimal('93.00'), Decimal('93.00'))]

We can see that the `Prev_year_pct_managed_drinking_water_services` column with the previous year's `Pct_managed_drinking_water_services` in the current year's row.  This shows that the lag was implemented correctly.

**NOTE:** The first year in every country partition has a `NULL` lag value since there is no previous year to extract a value from.

### 2. Determine the Annual Rate of Change between consecutive years.

Let’s go further and determine the Annual Rate of Change between consecutive years. That is, the difference between `Pct_managed_drinking_water_services` and `Prev_year_pct_managed_drinking_water_services`.

Therefore, we subtract the `previous percentage of managed drinking water` from the `current percentage of managed drinking water`.

The query will work similarly to the previous one, except that we reuse the LAG() function to calculate the Annual Rate of Change.

**NOTE:** SQL does not allow us to use the alias we had created within the same `SELECT` statement directly. Hence, we have to explicitly write the lag function again.

The results will then be returned as a new column named `ARC_pct_managed_drinking_water_services`.

In [16]:
%%sql

SELECT
    Country_name,
    Time_period,
    Pct_managed_drinking_water_services,
    LAG(Pct_managed_drinking_water_services) OVER( PARTITION BY Country_name 
    ORDER BY Time_period ASC) AS Prev_year_pct_managed_drinking_water_services,
    Pct_managed_drinking_water_services - LAG(Pct_managed_drinking_water_services) OVER( PARTITION BY Country_name 
    ORDER BY Time_period ASC) AS ARC_pct_managed_drinking_water_services
FROM 
    united_nations.Access_to_Basic_Services
LIMIT 80;

 * mysql+pymysql://root:***@localhost:3306/united_nations
80 rows affected.


[('Afghanistan', 2015, Decimal('67.00'), None, None),
 ('Afghanistan', 2016, Decimal('69.67'), Decimal('67.00'), Decimal('2.67')),
 ('Afghanistan', 2017, Decimal('72.33'), Decimal('69.67'), Decimal('2.66')),
 ('Afghanistan', 2018, Decimal('75.33'), Decimal('72.33'), Decimal('3.00')),
 ('Afghanistan', 2019, Decimal('78.00'), Decimal('75.33'), Decimal('2.67')),
 ('Afghanistan', 2020, Decimal('80.33'), Decimal('78.00'), Decimal('2.33')),
 ('Algeria', 2015, Decimal('92.00'), None, None),
 ('Algeria', 2016, Decimal('93.00'), Decimal('92.00'), Decimal('1.00')),
 ('Algeria', 2017, Decimal('93.00'), Decimal('93.00'), Decimal('0.00')),
 ('Algeria', 2018, Decimal('93.00'), Decimal('93.00'), Decimal('0.00')),
 ('Algeria', 2019, Decimal('93.33'), Decimal('93.00'), Decimal('0.33')),
 ('Algeria', 2020, Decimal('93.33'), Decimal('93.33'), Decimal('0.00')),
 ('American Samoa', 2015, Decimal('100.00'), None, None),
 ('American Samoa', 2016, Decimal('100.00'), Decimal('100.00'), Decimal('0.00')),
 ('American Samoa', 2017, Decimal('100.00'), Decimal('100.00'), Decimal('0.00')),
 ('American Samoa', 2018, Decimal('100.00'), Decimal('100.00'), Decimal('0.00')),
 ('American Samoa', 2019, Decimal('100.00'), Decimal('100.00'), Decimal('0.00')),
 ('American Samoa', 2020, Decimal('100.00'), Decimal('100.00'), Decimal('0.00')),
 ('Angola', 2015, Decimal('50.33'), None, None),
 ('Angola', 2016, Decimal('51.00'), Decimal('50.33'), Decimal('0.67')),
 ('Angola', 2017, Decimal('51.33'), Decimal('51.00'), Decimal('0.33')),
 ('Angola', 2018, Decimal('52.33'), Decimal('51.33'), Decimal('1.00')),
 ('Angola', 2019, Decimal('52.33'), Decimal('52.33'), Decimal('0.00')),
 ('Angola', 2020, Decimal('52.33'), Decimal('52.33'), Decimal('0.00')),
 ('Anguilla', 2015, Decimal('97.00'), None, None),
 ('Anguilla', 2016, Decimal('97.00'), Decimal('97.00'), Decimal('0.00')),
 ('Anguilla', 2017, Decimal('97.00'), Decimal('97.00'), Decimal('0.00')),
 ('Antigua and Barbuda', 2015, Decimal('97.00'), None, None),
 ('Antigua and Barbuda', 2016, Decimal('97.00'), Decimal('97.00'), Decimal('0.00')),
 ('Antigua and Barbuda', 2017, Decimal('97.00'), Decimal('97.00'), Decimal('0.00')),
 ('Argentina', 2015, Decimal('97.33'), None, None),
 ('Argentina', 2016, Decimal('97.33'), Decimal('97.33'), Decimal('0.00')),
 ('Argentina', 2017, Decimal('100.00'), Decimal('97.33'), Decimal('2.67')),
 ('Argentina', 2018, Decimal('100.00'), Decimal('100.00'), Decimal('0.00')),
 ('Argentina', 2019, Decimal('100.00'), Decimal('100.00'), Decimal('0.00')),
 ('Argentina', 2020, Decimal('100.00'), Decimal('100.00'), Decimal('0.00')),
 ('Armenia', 2015, Decimal('99.67'), None, None),
 ('Armenia', 2016, Decimal('99.67'), Decimal('99.67'), Decimal('0.00')),
 ('Armenia', 2017, Decimal('100.00'), Decimal('99.67'), Decimal('0.33')),
 ('Armenia', 2018, Decimal('100.00'), Decimal('100.00'), Decimal('0.00')),
 ('Armenia', 2019, Decimal('100.00'), Decimal('100.00'), Decimal('0.00')),
 ('Armenia', 2020, Decimal('100.00'), Decimal('100.00'), Decimal('0.00')),
 ('Aruba', 2015, Decimal('98.00'), None, None),
 ('Aruba', 2016, Decimal('98.00'), Decimal('98.00'), Decimal('0.00')),
 ('Australia', 2015, Decimal('100.00'), None, None),
 ('Australia', 2016, Decimal('100.00'), Decimal('100.00'), Decimal('0.00')),
 ('Australia', 2017, Decimal('100.00'), Decimal('100.00'), Decimal('0.00')),
 ('Australia', 2018, Decimal('100.00'), Decimal('100.00'), Decimal('0.00')),
 ('Australia', 2019, Decimal('100.00'), Decimal('100.00'), Decimal('0.00')),
 ('Australia', 2020, Decimal('100.00'), Decimal('100.00'), Decimal('0.00')),
 ('Azerbaijan', 2015, Decimal('91.67'), None, None),
 ('Azerbaijan', 2016, Decimal('92.33'), Decimal('91.67'), Decimal('0.66')),
 ('Azerbaijan', 2017, Decimal('93.67'), Decimal('92.33'), Decimal('1.34')),
 ('Azerbaijan', 2018, Decimal('94.33'), Decimal('93.67'), Decimal('0.66')),
 ('Azerbaijan', 2019, Decimal('95.00'), Decimal('94.33'), Decimal('0.67')),
 ('Azerbaijan', 2020, Decimal('95.67'), Decimal('95.00'), Decimal('0

There is a new column containing the **Annual Rate of Change** values calculated based on the `Pct_managed_drinking_water_services` column values and the `Prev_year_pct_managed_drinking_water_services` column values.

For instance, we can see that in Afghanistan, in the year 2017, the percentage of managed drinking water services increased to `72.33` from `69.67` in the previous year, resulting in an Annual Rate of Change of `2.66`.

In Bangladesh, the percentage of managed drinking water services in `2016` **remained the same** as the previous year, and therefore the Annual Rate of Change was `0`.

**NOTE:** A `None` value is returned for rows with `NULL` lag values as discussed earlier. 

## Summary

Further analysis can now be done to understand the factors influencing the change in the water access percentages over time.

#  

<div align="center" style=" font-size: 80%; text-align: center; margin: 0 auto">
<img src="https://raw.githubusercontent.com/Explore-AI/Pictures/master/ExploreAI_logos/EAI_Blue_Dark.png"  style="width:200px";/>
</div>